README.md
# note-duration

[![Code Climate](https://codeclimate.com/github/danigb/note-duration/badges/gpa.svg)](https://codeclimate.com/github/danigb/note-duration)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/feross/standard)


Parse note durations. It can convert from name to
[value](http://en.wikipedia.org/wiki/Note_value) and viceversa.

## Usage

Install the node module: `npm install --save node-duration`
and require it from you file:

```js
var duration = require('note-duration');
```

You can convert from names to values:

```js
duration('long');          // => 4
duration('double');         // => 2
duration('whole');          // => 1
duration('half');           // => 1/2
duration('quarter');        // => 1/4
duration('eighth');         // => 1/8
duration('sixteenth');      // => 1/16
duration('thirty-second');  // => 1/32
```

From letter and dots to values:

```js
duration("h");   // => 1/2
duration("h.");  // => dot: 1/2 + 1/4
duration("h.."); // => double dot: 1/2 + 1/4 + 1/8
duration("ht");  // => triplet: (1/2 + 1/ 2) / 3
duration("q");   // => 1/4
duration("q.");  // => dot: 1/4 + 1/8
duration("q.."); // => double dot: 1/4 + 1/8 + 1/16
duration("qt");  // => triplet: (1/4 + 1/4) / 3
```

From number string to value:

```js
duration("2");   // => 1/2
duration("2.");  // => dot: 1/2 + 1/4
duration("2t");  // => triplet: (1/2 + 1/ 2) / 3
duration("2.."); // => double dot: 1/2 + 1/4 + 1/8
duration("4");   // => 1/4
duration("4.");  // => dot: 1/4 + 1/8
duration("4.."); // => double dot: 1/4 + 1/8 + 1/16
duration("4t");  // => triplet: (1/4 + 1/4) / 3
```


The `duration.toString` method does the opposite: convert from value to letter and dots:

```js
duration.toString(1/2 + 1/4);        // => "h."
duration.toString(1/4 + 1/2 + 1/8);  // => "q.."
```

## License

MIT License


CHANGELOG
[MASTER]

[1.3.0]
- Add triplets
- Move to standard js style

[1.2.0]
- duration can be expressed in numbers (duration(4) == 0.25)

[1.1.1]
- duration(null) returns null
- duration.toString(null) returns null

[1.1.0]
- Change API: add toString method.

[1.0.0]
- First version


test/duration.js
var vows = require('vows')
var assert = require('assert')

var duration = require('..')

vows.describe('time').addBatch({
  'null as parameter': function () {
    assert.equal(duration(null), null)
    assert.equal(duration.toString(null), null)
  },
  'value from name': function () {
    assert.equal(duration('long'), 4)
    assert.equal(duration('double'), 2)
    assert.equal(duration('whole'), 1)
    assert.equal(duration('half'), 1 / 2)
    assert.equal(duration('quarter'), 1 / 4)
    assert.equal(duration('eighth'), 1 / 8)
    assert.equal(duration('sixteenth'), 1 / 16)
    assert.equal(duration('thirty-second'), 1 / 32)
  },
  'value from symbol': function () {
    assert.equal(duration('l'), 4)
    assert.equal(duration('l.'), 4 + 2)
    assert.equal(duration('l..'), 4 + 2 + 1)
    assert.equal(duration('d'), 2)
    assert.equal(duration('d.'), 2 + 1)
    assert.equal(duration('d..'), 2 + 1 + 1 / 2)
    assert.equal(duration('w'), 1)
    assert.equal(duration('w.'), 1 + 1 / 2)
    assert.equal(duration('w..'), 1 + 1 / 2 + 1 / 4)
    assert.equal(duration('h'), 1 / 2)
    assert.equal(duration('h.'), 1 / 2 + 1 / 4)
    assert.equal(duration('h..'), 1 / 2 + 1 / 4 + 1 / 8)
    assert.equal(duration('q'), 1 / 4)
    assert.equal(duration('q.'), 1 / 4 + 1 / 8)
    assert.equal(duration('q..'), 1 / 4 + 1 / 8 + 1 / 16)
    assert.equal(duration('e'), 1 / 8)
    assert.equal(duration('e.'), 1 / 8 + 1 / 16)
    assert.equal(duration('e..'), 1 / 8 + 1 / 16 + 1 / 32)
    assert.equal(duration('s'), 1 / 16)
    assert.equal(duration('s.'), 1 / 16 + 1 / 32)
    assert.equal(duration('s..'), 1 / 16 + 1 / 32 + 1 / 64)
  },
  'value from number': function () {
    assert.equal(duration('0.25'), 4)
    assert.equal(duration('0.25.'), 4 + 2)
    assert.equal(duration('0.25..'), 4 + 2 + 1)
    assert.equal(duration('0.5'), 2)
    assert.equal(duration('0.5.'), 2 + 1)
    assert.equal(duration('0.5..'), 2 + 1 + 1 / 2)
    assert.equal(duration('1'), 1)
    assert.equal(duration('1.'), 1 + 1 / 2)
    assert.equal(duration('1..'), 1 + 1 / 2 + 1 / 4)
    assert.equal(duration('2'), 1 / 2)
    assert.equal(duration('2.'), 1 / 2 + 1 / 4)
    assert.equal(duration('2..'), 1 / 2 + 1 / 4 + 1 / 8)
    assert.equal(duration('4'), 1 / 4)
    assert.equal(duration('4.'), 1 / 4 + 1 / 8)
    assert.equal(duration('4..'), 1 / 4 + 1 / 8 + 1 / 16)
    assert.equal(duration('8'), 1 / 8)
    assert.equal(duration('8.'), 1 / 8 + 1 / 16)
    assert.equal(duration('8..'), 1 / 8 + 1 / 16 + 1 / 32)
    assert.equal(duration('16'), 1 / 16)
    assert.equal(duration('16.'), 1 / 16 + 1 / 32)
    assert.equal(duration('16..'), 1 / 16 + 1 / 32 + 1 / 64)
  },
  'triplets': function() {
    assert.equal(duration('qt'), (1 / 4 + 1 / 4 ) / 3);
  },
  'toString': function () {
    assert.equal(duration.toString(4), 'l')
    assert.equal(duration.toString(4 + 2), 'l.')
    assert.equal(duration.toString(4 + 2 + 1), 'l..')
    assert.equal(duration.toString(2), 'd')
    assert.equal(duration.toString(2 + 1), 'd.')
    assert.equal(duration.toString(2 + 1 + 1 / 2), 'd..')
    assert.equal(duration.toString(1), 'w')
    assert.equal(duration.toString(1 + 1 / 2), 'w.')
    assert.equal(duration.toString(1 + 1 / 2 + 1 / 4), 'w..')
    assert.equal(duration.toString(1 / 2), 'h')
    assert.equal(duration.toString(1 / 2 + 1 / 4), 'h.')
    assert.equal(duration.toString(1 / 2 + 1 / 4 + 1 / 8), 'h..')
    assert.equal(duration.toString(1 / 4), 'q')
    assert.equal(duration.toString(1 / 4 + 1 / 8), 'q.')
    assert.equal(duration.toString(1 / 4 + 1 / 8 + 1 / 16), 'q..')
    assert.equal(duration.toString(1 / 8), 'e')
    assert.equal(duration.toString(1 / 8 + 1 / 16), 'e.')
    assert.equal(duration.toString(1 / 8 + 1 / 16 + 1 / 32), 'e..')
    assert.equal(duration.toString(1 / 16), 's')
    assert.equal(duration.toString(1 / 16 + 1 / 32), 's.')
    assert.equal(duration.toString(1 / 16 + 1 / 32 + 1 / 64), 's..')
  }
}).export(module)


index.js
'use strict'

var names = ['long', 'double', 'whole', 'half', 'quarter', 'eighth', 'sixteenth', 'thirty-second']
var values = [4, 2, 1, 1 / 2, 1 / 4, 1 / 8, 1 / 16, 1 / 32]

var namesToValues = {}
for (var i = 0; i < values.length; i++) {
  var name = names[i]
  var value = values[i]
  var short = name[0]
  var num = '' + (1 / value)
  namesToValues[name] = value
  namesToValues[short] = namesToValues[num] = value
  namesToValues[short + '.'] = namesToValues[num + '.'] = value + value / 2
  namesToValues[short + '..'] = namesToValues[num + '..'] = value + value / 2 + value / 4
  namesToValues[short + 't'] = namesToValues[num + 't'] = (value + value) / 3
}

var valuesToNames = {}
names.forEach(function (name, index) {
  var value = values[index]
  valuesToNames['' + value] = name[0]
  valuesToNames['' + (value + value / 2)] = name[0] + '.'
  valuesToNames['' + (value + value / 2 + value / 4)] = name[0] + '..'
})

var duration = function (name) {
  return namesToValues['' + name]
}

duration.toString = function (value) {
  return valuesToNames['' + value]
}

module.exports = duration
